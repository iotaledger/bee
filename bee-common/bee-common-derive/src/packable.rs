// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use proc_macro2::TokenStream;
use proc_macro_error::abort;
use quote::{format_ident, quote};
use syn::{
    parse::{Parse, ParseStream},
    spanned::Spanned,
    Attribute, Fields, FieldsNamed, FieldsUnnamed, Generics, Ident, Index, Token, Type, TypePath, Variant,
};

pub(crate) fn gen_struct_bodies(struct_fields: Fields) -> (TokenStream, TokenStream, TokenStream) {
    let (pack, unpack, packed_len);

    match struct_fields {
        Fields::Named(FieldsNamed { named, .. }) => {
            let (labels, types): (Vec<&Ident>, Vec<&Type>) = named
                .iter()
                // This is a named field, which means its `ident` cannot be `None`
                .map(|field| (field.ident.as_ref().unwrap(), &field.ty))
                .unzip();

            pack = quote! {
                #(<#types>::pack(&self.#labels, packer)?;) *
                Ok(())
            };

            unpack = quote! {
                Ok(Self { #(#labels: <#types>::unpack(unpacker).map_err(bee_common::packable::UnpackError::coerce)?,)* })
            };

            packed_len = quote! {
                0 #(+ <#types>::packed_len(&self.#labels)) *
            }
        }
        Fields::Unnamed(FieldsUnnamed { unnamed, .. }) => {
            let (indices, types): (Vec<Index>, Vec<&Type>) = unnamed
                .iter()
                .enumerate()
                .map(|(index, field)| (index.into(), &field.ty))
                .unzip();

            pack = quote! {
                #(<#types>::pack(&self.#indices, packer)?;) *
                Ok(())
            };

            unpack = quote! {
                Ok(Self(#(<#types>::unpack(unpacker).map_err(bee_common::packable::UnpackError::coerce)?), *))
            };

            packed_len = quote! {
                0 #(+ <#types>::packed_len(&self.#indices)) *
            }
        }
        Fields::Unit => {
            pack = quote!(Ok(()));
            unpack = quote!(Ok(Self));
            packed_len = quote!(0);
        }
    }

    (pack, unpack, packed_len)
}

pub(crate) fn gen_enum_bodies<'a>(
    variants: impl Iterator<Item = &'a Variant> + 'a,
    ty: Type,
) -> (TokenStream, TokenStream, TokenStream) {
    const TYPES: &[&str] = &["u8", "u16", "u32", "u64", "i8", "i16", "i32", "i64"];

    match &ty {
        Type::Path(TypePath { path, .. }) if TYPES.iter().any(|ty| path.is_ident(ty)) => (),
        _ => abort!(ty.span(), "Tags for enums can only be sized integers."),
    }

    let mut indices = Vec::<(Index, &Ident)>::new();

    let mut pack_branches = Vec::new();
    let mut unpack_branches = Vec::new();
    let mut packed_len_branches = Vec::new();

    for variant in variants {
        let Variant {
            attrs, ident, fields, ..
        } = variant;

        let curr_index = parse_attr::<Index>("tag", attrs).unwrap_or_else(|| {
            abort!(
                ident.span(),
                "All variants of an enum that derives `Packable` require a `#[packable(tag = ...)]` attribute."
            )
        });

        match indices.binary_search_by(|(index, _)| index.index.cmp(&curr_index.index)) {
            Ok(pos) => abort!(
                curr_index.span,
                "The tag `{}` is already being used for the `{}` variant.",
                curr_index.index,
                indices[pos].1
            ),
            Err(pos) => indices.insert(pos, (curr_index.clone(), ident)),
        }

        let tag = proc_macro2::Literal::u64_unsuffixed(curr_index.index as u64);

        let pack_branch: TokenStream;
        let unpack_branch: TokenStream;
        let packed_len_branch: TokenStream;

        match fields {
            Fields::Named(FieldsNamed { named, .. }) => {
                let (labels, types): (Vec<&Ident>, Vec<&Type>) = named
                    .iter()
                    // This is a named field, which means its `ident` cannot be `None`
                    .map(|field| (field.ident.as_ref().unwrap(), &field.ty))
                    .unzip();

                pack_branch = quote! {
                    Self::#ident{#(#labels), *} => {
                        (#tag as #ty).pack(packer)?;
                        #(<#types>::pack(&#labels, packer)?;) *
                        Ok(())
                    }
                };

                unpack_branch = quote! {
                    #tag => Ok(Self::#ident{
                        #(#labels: <#types>::unpack(unpacker).map_err(bee_common::packable::UnpackError::coerce)?,) *
                    })
                };

                packed_len_branch = quote! {
                    Self::#ident{#(#labels), *} => {
                        (#tag as #ty).packed_len() #(+ <#types>::packed_len(&#labels)) *
                    }
                };
            }
            Fields::Unnamed(FieldsUnnamed { unnamed, .. }) => {
                let (fields, types): (Vec<Ident>, Vec<&Type>) = unnamed
                    .iter()
                    .enumerate()
                    .map(|(index, field)| (format_ident!("field_{}", index), &field.ty))
                    .unzip();

                pack_branch = quote! {
                    Self::#ident(#(#fields), *) => {
                        (#tag as #ty).pack(packer)?;
                        #(<#types>::pack(&#fields, packer)?;) *
                        Ok(())
                    }
                };

                unpack_branch = quote! {
                    #tag => Ok(Self::#ident(
                        #(<#types>::unpack(unpacker).map_err(bee_common::packable::UnpackError::coerce)?), *
                    ))
                };

                packed_len_branch = quote! {
                    Self::#ident(#(#fields), *) => {
                        (#tag as #ty).packed_len() #(+ <#types>::packed_len(&#fields)) *
                    }
                };
            }
            Fields::Unit => {
                pack_branch = quote! {
                    Self::#ident => (#tag as #ty).pack(packer)
                };

                unpack_branch = quote! {
                    #tag => Ok(Self::#ident)
                };

                packed_len_branch = quote! {
                    Self::#ident => (#tag as #ty).packed_len()
                };
            }
        };
        pack_branches.push(pack_branch);
        unpack_branches.push(unpack_branch);
        packed_len_branches.push(packed_len_branch);
    }

    (
        quote! {
            match self {
                #(#pack_branches,) *
            }
        },
        quote! {
            match unpacker.unpack_infallible::<#ty>()? {
                #(#unpack_branches,) *
                id => Err(bee_common::packable::UnpackError::Packable(bee_common::packable::UnknownTagError(id).into()))
            }
        },
        quote! {
            match self {
                #(#packed_len_branches,) *
            }
        },
    )
}

pub(crate) fn gen_impl(
    ident: &Ident,
    generics: &Generics,
    error_type: TokenStream,
    pack_body: TokenStream,
    unpack_body: TokenStream,
    packed_len_body: TokenStream,
) -> TokenStream {
    let params = generics.params.iter();

    quote! {
        impl <#(#params: Packable,) *> Packable for #ident #generics {
            type Error = #error_type;

            fn pack<P: bee_common::packable::Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
                #pack_body
            }

            fn unpack<U: bee_common::packable::Unpacker>(unpacker: &mut U) -> Result<Self, bee_common::packable::UnpackError<Self::Error, U::Error>> {
                #unpack_body
            }

            fn packed_len(&self) -> usize {
                #packed_len_body
            }
        }
    }
}

pub(crate) fn parse_attr<T: Parse>(ident: &str, attrs: &[Attribute]) -> Option<T> {
    struct AttrArg<T> {
        ident: Ident,
        value: T,
    }

    impl<T: Parse> Parse for AttrArg<T> {
        fn parse(input: ParseStream) -> syn::Result<Self> {
            let ident = input.parse::<Ident>()?;
            let _ = input.parse::<Token![=]>()?;
            let value = input.parse::<T>()?;

            Ok(Self { ident, value })
        }
    }

    for attr in attrs {
        if attr.path.is_ident("packable") {
            match attr.parse_args::<AttrArg<T>>() {
                Ok(arg) if arg.ident == ident => return Some(arg.value),
                _ => (),
            }
        }
    }

    None
}
