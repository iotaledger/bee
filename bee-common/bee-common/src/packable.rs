// Copyright 2020-2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! A module that provides a `Packable` trait to serialize and deserialize types.

pub use std::io::{Read, Write};

/// A trait to pack and unpack types to and from bytes.
pub trait Packable {
    /// Associated error type.
    type Error: std::fmt::Debug;

    /// Returns the length of the packed bytes.
    fn packed_len(&self) -> usize;

    /// Packs the instance to bytes and writes them to the passed writer.
    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error>;

    /// Packs the instance to bytes and writes them to a newly allocated vector.
    fn pack_new(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(self.packed_len());
        // Packing to bytes can't fail.
        self.pack(&mut bytes).unwrap();

        bytes
    }

    /// Reads bytes from the passed reader and unpacks them into an instance.
    fn unpack_inner<R: Read + ?Sized, const TRUSTED: bool>(reader: &mut R) -> Result<Self, Self::Error>
    where
        Self: Sized;

    /// Reads untrusted bytes from the passed reader and unpacks them into an instance.
    fn unpack<R: Read + ?Sized>(reader: &mut R) -> Result<Self, Self::Error>
    where
        Self: Sized,
    {
        Self::unpack_inner::<R, false>(reader)
    }
}

impl<const N: usize> Packable for [u8; N] {
    type Error = std::io::Error;

    fn packed_len(&self) -> usize {
        N
    }

    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
        writer.write_all(self)?;

        Ok(())
    }

    fn unpack_inner<R: Read + ?Sized, const TRUSTED: bool>(reader: &mut R) -> Result<Self, Self::Error>
    where
        Self: Sized,
    {
        let mut bytes = [0u8; N];
        reader.read_exact(&mut bytes)?;

        Ok(bytes)
    }
}

impl Packable for bool {
    type Error = std::io::Error;

    fn packed_len(&self) -> usize {
        (*self as u8).packed_len()
    }

    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
        (*self as u8).pack(writer)
    }

    fn unpack_inner<R: Read + ?Sized, const TRUSTED: bool>(reader: &mut R) -> Result<Self, Self::Error>
    where
        Self: Sized,
    {
        Ok(!matches!(u8::unpack_inner::<R, TRUSTED>(reader)?, 0))
    }
}

impl<P> Packable for Vec<P>
where
    P: Packable,
    P::Error: From<std::io::Error>,
{
    type Error = P::Error;

    fn packed_len(&self) -> usize {
        0u64.packed_len() + self.iter().map(Packable::packed_len).sum::<usize>()
    }

    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
        (self.len() as u64).pack(writer)?;
        self.iter().try_for_each(|x| x.pack(writer))
    }

    fn unpack_inner<R: Read + ?Sized, const TRUSTED: bool>(reader: &mut R) -> Result<Self, Self::Error>
    where
        Self: Sized,
    {
        (0..u64::unpack_inner::<R, TRUSTED>(reader)?)
            .map(|_| P::unpack_inner::<R, TRUSTED>(reader))
            .collect()
    }
}

/// Error that occurs on `Option<P: Packable>` operations.
#[derive(Debug)]
pub enum OptionError<E> {
    /// Error that occurs on boolean `Packable` operations.
    Bool(<bool as Packable>::Error),
    /// Error that occurs on inner `Packable` operations.
    Inner(E),
}

impl<E> From<E> for OptionError<E> {
    fn from(inner: E) -> Self {
        OptionError::Inner(inner)
    }
}

impl<P: Packable> Packable for Option<P> {
    type Error = OptionError<P::Error>;

    fn packed_len(&self) -> usize {
        true.packed_len() + self.as_ref().map_or(0, Packable::packed_len)
    }

    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
        if let Some(p) = self {
            true.pack(writer).map_err(OptionError::Bool)?;
            p.pack(writer).map_err(OptionError::Inner)?;
        } else {
            false.pack(writer).map_err(OptionError::Bool)?;
        }

        Ok(())
    }

    fn unpack_inner<R: Read + ?Sized, const TRUSTED: bool>(reader: &mut R) -> Result<Self, Self::Error>
    where
        Self: Sized,
    {
        Ok(
            if bool::unpack_inner::<R, TRUSTED>(reader).map_err(OptionError::Bool)? {
                Some(P::unpack_inner::<R, TRUSTED>(reader).map_err(OptionError::Inner)?)
            } else {
                None
            },
        )
    }
}

macro_rules! impl_packable_for_num {
    ($ty:ident) => {
        impl Packable for $ty {
            type Error = std::io::Error;

            fn packed_len(&self) -> usize {
                std::mem::size_of_val(&self.to_le_bytes())
            }

            fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
                writer.write_all(&self.to_le_bytes())?;

                Ok(())
            }

            fn unpack_inner<R: Read + ?Sized, const TRUSTED: bool>(reader: &mut R) -> Result<Self, Self::Error>
            where
                Self: Sized,
            {
                let mut bytes = [0; $ty::MIN.to_le_bytes().len()];
                reader.read_exact(&mut bytes)?;

                Ok($ty::from_le_bytes(bytes))
            }
        }
    };
}

impl_packable_for_num!(i8);
impl_packable_for_num!(u8);
impl_packable_for_num!(i16);
impl_packable_for_num!(u16);
impl_packable_for_num!(i32);
impl_packable_for_num!(u32);
impl_packable_for_num!(i64);
impl_packable_for_num!(u64);
#[cfg(has_i128)]
impl_packable_for_num!(i128);
#[cfg(has_u128)]
impl_packable_for_num!(u128);
