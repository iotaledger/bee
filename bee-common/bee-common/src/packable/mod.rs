// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! A module that provides a `Packable` trait to serialize and deserialize types.

extern crate alloc;

mod error;
#[cfg(feature = "io")]
mod io;
mod packer;
mod unpacker;

use alloc::{boxed::Box, vec::Vec};
use core::convert::Infallible;

pub use error::{UnknownTagError, UnpackError};
pub use packer::{Packer, VecPacker};
pub use unpacker::{SliceUnpacker, UnexpectedEOF, Unpacker};

pub use bee_common_derive::Packable;

/// A type that can be packed and unpacked.
///
/// Almost all basic sized types implement this trait. This trait can be derived using the
/// `bee_common_derive::Packable` macro. If you need to implement this trait manually, use the provided
/// implementations as a guide.
pub trait Packable: Sized {
    /// The error type that can be returned if some semantic error occurs while unpacking.
    ///
    /// It is recommended to use `core::convert::Infallible` if this kind of error cannot happen or
    /// `UnknownTagError` when implementing this trait for an enum.
    type Error;
    /// Pack this value into the given `Packer`.
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
    /// The size of the value in bytes after being packed.
    fn packed_len(&self) -> usize;
    /// Unpack this value from the given `Unpacker`.
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>>;
}

macro_rules! impl_packable_for_int {
    ($ty:ty) => {
        impl Packable for $ty {
            type Error = Infallible;

            fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
                packer.pack_bytes(&self.to_le_bytes())
            }

            fn packed_len(&self) -> usize {
                core::mem::size_of::<Self>()
            }

            fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
                let mut bytes = [0u8; core::mem::size_of::<Self>()];
                unpacker.unpack_bytes(&mut bytes)?;
                Ok(Self::from_le_bytes(bytes))
            }
        }
    };
}

impl_packable_for_int!(u8);
impl_packable_for_int!(u16);
impl_packable_for_int!(u32);
impl_packable_for_int!(u64);
#[cfg(has_u128)]
impl_packable_for_int!(u128);

/// `usize` integers are packed and unpacked as `u64` integers according to the spec.
impl Packable for usize {
    type Error = Infallible;

    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (*self as u64).pack(packer)
    }

    fn packed_len(&self) -> usize {
        core::mem::size_of::<u64>()
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        Ok(unpacker.unpack_infallible::<u64>().map_err(UnpackError::Unpacker)? as usize)
    }
}

impl_packable_for_int!(i8);
impl_packable_for_int!(i16);
impl_packable_for_int!(i32);
impl_packable_for_int!(i64);
#[cfg(has_i128)]
impl_packable_for_int!(i128);

/// `isize` integers are packed and unpacked as `i64` integers according to the spec.
impl Packable for isize {
    type Error = Infallible;

    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (*self as i64).pack(packer)
    }

    fn packed_len(&self) -> usize {
        core::mem::size_of::<i64>()
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        Ok(unpacker.unpack_infallible::<i64>().map_err(UnpackError::Unpacker)? as isize)
    }
}

impl Packable for bool {
    type Error = Infallible;

    /// Booleans are packed as `u8` integers following Rust's data layout.
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (*self as u8).pack(packer)
    }

    fn packed_len(&self) -> usize {
        core::mem::size_of::<u8>()
    }

    /// Booleans are unpacked if the byte used to represent them is non-zero.
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        Ok(unpacker.unpack_infallible::<u8>().map_err(UnpackError::Unpacker)? != 0)
    }
}

/// Error type raised when a semantic error occurs while unpacking an option.
#[derive(Debug)]
pub enum UnpackOptionError<T: Packable> {
    /// The tag found while unpacking is not valid.
    UnknownTag(u8),
    /// A semantic error for the underlying type was raised.
    Inner(T::Error),
}

/// Options are packed and unpacked using `0u8` as the prefix for `None` and `1u8` as the prefix
/// for `Some`.
impl<T: Packable> Packable for Option<T> {
    type Error = UnpackOptionError<T>;

    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        match self {
            None => 0u8.pack(packer),
            Some(item) => {
                1u8.pack(packer)?;
                item.pack(packer)
            }
        }
    }

    fn packed_len(&self) -> usize {
        0u8.packed_len()
            + match self {
                Some(item) => item.packed_len(),
                None => 0,
            }
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        match unpacker.unpack_infallible::<u8>()? {
            0 => Ok(None),
            1 => Ok(Some(
                T::unpack(unpacker).map_err(|err| err.map(UnpackOptionError::Inner))?,
            )),
            n => Err(UnpackError::Packable(Self::Error::UnknownTag(n))),
        }
    }
}

impl<T: Packable, const N: usize> Packable for [T; N] {
    type Error = T::Error;

    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        for item in self.iter() {
            item.pack(packer)?;
        }

        Ok(())
    }

    fn packed_len(&self) -> usize {
        self.iter().map(T::packed_len).sum::<usize>()
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        use core::mem::MaybeUninit;

        // Safety: an uninitialized array of `MaybeUninit`s is safe to be considered initialized.
        // FIXME: replace with [`uninit_array`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.uninit_array)
        // when stabilized.
        let mut array = unsafe { MaybeUninit::<[MaybeUninit<T>; N]>::uninit().assume_init() };

        for item in array.iter_mut() {
            let unpacked = T::unpack(unpacker)?;
            // Safety: each `item` is only visited once so we are never overwriting nor dropping
            // values that are already initialized.
            unsafe {
                item.as_mut_ptr().write(unpacked);
            }
        }

        // Safety: We traversed the whole array and initialized every item.
        // FIXME: replace with [`array_assume_init`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.array_assume_init)
        // when stabilized.
        Ok(unsafe { (&array as *const [MaybeUninit<T>; N] as *const Self).read() })
    }
}

impl<T: Packable> Packable for Vec<T> {
    type Error = T::Error;

    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        // The length of any dynamically-sized sequence must be prefixed.
        self.len().pack(packer)?;

        for item in self.iter() {
            item.pack(packer)?;
        }

        Ok(())
    }

    fn packed_len(&self) -> usize {
        0usize.packed_len() + self.iter().map(T::packed_len).sum::<usize>()
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        // The length of any dynamically-sized sequence must be prefixed.
        let len = unpacker.unpack_infallible::<usize>().map_err(UnpackError::Unpacker)?;

        let mut vec = Self::with_capacity(len);

        for _ in 0..len {
            let item = T::unpack(unpacker)?;
            vec.push(item);
        }

        Ok(vec)
    }
}

impl<T: Packable> Packable for Box<[T]> {
    type Error = T::Error;

    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        // The length of any dynamically-sized sequence must be prefixed.
        self.len().pack(packer)?;

        for item in self.iter() {
            item.pack(packer)?;
        }

        Ok(())
    }

    fn packed_len(&self) -> usize {
        0usize.packed_len() + self.iter().map(T::packed_len).sum::<usize>()
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>> {
        Ok(Vec::<T>::unpack(unpacker)?.into_boxed_slice())
    }
}
