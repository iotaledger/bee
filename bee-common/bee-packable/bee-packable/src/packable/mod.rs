// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! A module that provides a `Packable` trait to serialize and deserialize types.

extern crate alloc;

mod array;
mod bool;
mod r#box;
mod integer;
mod option;
mod vec;
mod vec_prefix;

pub use option::UnpackOptionError;
pub use vec_prefix::VecPrefix;

pub use crate::{
    error::{PackError, UnknownTagError, UnpackError},
    packer::{Packer, VecPacker},
    unpacker::{SliceUnpacker, UnexpectedEOF, Unpacker},
};

pub use bee_packable_derive::Packable;

use alloc::vec::Vec;
use core::convert::AsRef;

/// A type that can be packed and unpacked.
///
/// Almost all basic sized types implement this trait. This trait can be derived using the
/// `bee_common_derive::Packable` macro. If you need to implement this trait manually, use the provided
/// implementations as a guide.
pub trait Packable: Sized {
    /// The error type that can be returned if some semantic error occurs while packing.
    ///
    /// It is recommended to use `core::convert::Infallible` if this kind of error cannot happen or
    /// `UnknownTagError` when implementing this trait for an enum.
    type PackError;

    /// The error type that can be returned if some semantic error occurs while unpacking.
    ///
    /// It is recommended to use `core::convert::Infallible` if this kind of error cannot happen or
    /// `UnknownTagError` when implementing this trait for an enum.
    type UnpackError;

    /// Packs this value into the given `Packer`.
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), PackError<Self::PackError, P::Error>>;

    /// The size of the value in bytes after being packed.
    fn packed_len(&self) -> usize;

    /// Convenience method that packs this value into a `Vec<u8>`.
    fn pack_to_vec(&self) -> Result<Vec<u8>, Self::PackError> {
        let mut packer = VecPacker::with_capacity(self.packed_len());

        // Packing to bytes will not fail but packing the value itself might.
        self.pack(&mut packer).map_err(|err| match err {
            PackError::Packable(err) => err,
            PackError::Packer(err) => match err {},
        })?;

        Ok(packer.into_vec())
    }

    /// Unpacks this value from the given `Unpacker`.
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::UnpackError, U::Error>>;

    /// Unpacks this value from a type that implements `AsRef<[u8]>`.
    fn unpack_from_slice<T: AsRef<[u8]>>(bytes: T) -> Result<Self, UnpackError<Self::UnpackError, UnexpectedEOF>> {
        let mut unpacker = SliceUnpacker::new(bytes.as_ref());
        Packable::unpack(&mut unpacker)
    }
}
