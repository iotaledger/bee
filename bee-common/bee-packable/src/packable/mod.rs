// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! A module that provides a `Packable` trait to serialize and deserialize types.

mod array;
mod bool;
mod r#box;
mod integer;
mod option;
mod vec;
mod vec_prefix;

pub use option::UnpackOptionError;
pub use vec_prefix::VecPrefix;

pub use crate::{
    error::{UnknownTagError, UnpackError},
    packer::{Packer, VecPacker},
    unpacker::{SliceUnpacker, UnexpectedEOF, Unpacker},
};

pub use bee_packable_derive::Packable;

/// A type that can be packed and unpacked.
///
/// Almost all basic sized types implement this trait. This trait can be derived using the
/// `bee_common_derive::Packable` macro. If you need to implement this trait manually, use the provided
/// implementations as a guide.
pub trait Packable: Sized {
    /// The error type that can be returned if some semantic error occurs while unpacking.
    ///
    /// It is recommended to use `core::convert::Infallible` if this kind of error cannot happen or
    /// `UnknownTagError` when implementing this trait for an enum.
    type Error;

    /// Pack this value into the given `Packer`.
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
    /// The size of the value in bytes after being packed.

    fn packed_len(&self) -> usize;

    /// Unpack this value from the given `Unpacker`.
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, UnpackError<Self::Error, U::Error>>;
}
