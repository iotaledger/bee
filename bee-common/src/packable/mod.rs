// Copyright 2020-2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! A module that provides a `Packable` trait to serialize and deserialize types.

mod packer;
mod unpacker;

pub use packer::{PackError, Packer};
pub use unpacker::{UnpackError, Unpacker};

/// A type that can be packed and unpacked.
///
/// Almost all basic sized types implement this trait. This trait can be derived using the
/// `bee_common_derive::Packable` macro. If you need to implement this trait manually, use the provided
/// implementations as a guide.
pub trait Packable: Sized {
    /// Pack this value into the given `Packer`.
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
    /// Unpack this value from the given `Unpacker`.
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error>;
}

macro_rules! impl_packable_for_int {
    ($ty:ty) => {
        impl Packable for $ty {
            fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
                packer.pack_bytes(&self.to_le_bytes())
            }

            fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
                let bytes: &[u8; core::mem::size_of::<Self>()] = unpacker.unpack_exact_bytes()?;
                Ok(Self::from_le_bytes(*bytes))
            }
        }
    };
}

impl_packable_for_int!(u8);
impl_packable_for_int!(u16);
impl_packable_for_int!(u32);
impl_packable_for_int!(u64);
#[cfg(has_u128)]
impl_packable_for_int!(u128);

/// `usize` integers are packed and unpacked as `u64` integers according to the spec.
impl Packable for usize {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (*self as u64).pack(packer)
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        Ok(u64::unpack(unpacker)? as usize)
    }
}

impl_packable_for_int!(i8);
impl_packable_for_int!(i16);
impl_packable_for_int!(i32);
impl_packable_for_int!(i64);
#[cfg(has_i128)]
impl_packable_for_int!(i128);

/// `isize` integers are packed and unpacked as `i64` integers according to the spec.
impl Packable for isize {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (*self as i64).pack(packer)
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        Ok(u64::unpack(unpacker)? as isize)
    }
}

impl Packable for bool {
    /// Booleans are packed as `u8` integers following Rust's data layout.
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (*self as u8).pack(packer)
    }

    /// Booleans are unpacked if the byte used to represent them is non-zero.
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        Ok(u8::unpack(unpacker)? != 0)
    }
}

#[cfg(feature = "alloc")]
mod alloc_support {
    extern crate alloc;

    use super::{Packable, Packer, Unpacker};

    impl<T: Packable> Packable for alloc::vec::Vec<T> {
        fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
            // The length of any dynamically-sized sequence must be prefixed.
            self.len().pack(packer)?;

            for item in self.iter() {
                item.pack(packer)?;
            }

            Ok(())
        }

        fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
            // The length of any dynamically-sized sequence must be prefixed.
            let len = usize::unpack(unpacker)?;

            let mut vec = Self::with_capacity(len);

            for _ in 0..len {
                let item = T::unpack(unpacker)?;
                vec.push(item);
            }

            Ok(vec)
        }
    }

    impl<T: Packable> Packable for alloc::boxed::Box<[T]> {
        fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
            // The length of any dynamically-sized sequence must be prefixed.
            self.len().pack(packer)?;

            for item in self.iter() {
                item.pack(packer)?;
            }

            Ok(())
        }

        fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
            Ok(alloc::vec::Vec::<T>::unpack(unpacker)?.into_boxed_slice())
        }
    }
}

impl<T: Packable, const N: usize> Packable for [T; N] {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        for item in self.iter() {
            item.pack(packer)?;
        }

        Ok(())
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        use core::mem::MaybeUninit;

        // Safety: an uninitialized array of unitialized stuff is actually initialized.
        let mut array = unsafe { MaybeUninit::<[MaybeUninit<T>; N]>::uninit().assume_init() };

        for item in array.iter_mut() {
            let unpacked = T::unpack(unpacker)?;
            // Safety: each `item` is only visited once so we are never overwritting values that
            // are already initialized.
            unsafe {
                item.as_mut_ptr().write(unpacked);
            }
        }

        // Safety: We traversed the whole array and initialized every item.
        Ok(unsafe { (&array as *const [MaybeUninit<T>; N] as *const Self).read() })
    }
}

/// Options are packed and unpacked using `0u8` as the prefix for `None` and `1u8` as the prefix
/// for `Some`.
impl<T: Packable> Packable for Option<T> {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        match self {
            None => 0u8.pack(packer),
            Some(item) => {
                1u8.pack(packer)?;
                item.pack(packer)
            }
        }
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        match u8::unpack(unpacker)? {
            0 => Ok(None),
            1 => Ok(Some(T::unpack(unpacker)?)),
            n => Err(U::Error::invalid_variant::<Self>(n as u64)),
        }
    }
}
