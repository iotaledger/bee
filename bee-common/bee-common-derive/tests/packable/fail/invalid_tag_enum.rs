// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

#![allow(unused_imports)]

use bee_common::packable::Packable;
use bee_common_derive::Packable;

use core::convert::Infallible;

#[derive(Packable)]
#[packable(tag_ty = u8)]
#[packable(error = Infallible)]
pub enum OptI32 {
    #[packable(tag = [0; 32])]
    None,
    #[packable(tag = 1)]
    Some(i32)
}

fn main() {}
