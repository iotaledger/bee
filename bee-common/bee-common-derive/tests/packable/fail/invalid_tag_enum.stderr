error: All variants of an enum that derives `Packable` require a `#[packable(tag = ...)]` attribute.
  --> $DIR/invalid_tag_enum.rs:13:5
   |
13 |     None,
   |     ^^^^
